### PR TITLE
Don't require models to support more than 1 batch axis

### PR DESCRIPTION
**Pull request by inailuig**
when using the jax operators.
Somehow we missed this in https://github.com/netket/netket/pull/1506

Thanks to @jwnys for pointing it out.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Can you add a quick test?

File: netket/vqs/mc/kernels.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def local_value_kernel_jax(
     """
     σp, mel = O.get_conn_padded(σ)
     logpsi_σ = logpsi(pars, σ)
-    logpsi_σp = logpsi(pars, σp)
+    logpsi_σp = logpsi(pars, σp.reshape(-1, σp.shape[-1])).reshape(σp.shape[:-1])
```
Comment (gcarleo) [inline netket/vqs/mc/kernels.py:71]: https://chat.openai.com/share/eefd3861-16f7-4850-ac63-998784264c7b

Comment (inailuig) [reply netket/vqs/mc/kernels.py:71]: not really, there is a call to logpsi in between

Comment (jwnys) [reply netket/vqs/mc/kernels.py:71]: You should reshape also the \sigma, not only the \sigmap

Comment (gcarleo) [reply netket/vqs/mc/kernels.py:71]: For the sake of readability can we define an auxiliary small function that does this reshaping? Something like 'at_worst1d' (similar to atleast1d)

Comment (inailuig) [reply netket/vqs/mc/kernels.py:71]: @jwnys the chains should already be flattened [here](https://github.com/netket/netket/blob/fe7629b996250b0661e78fb5f79e83648d1bc3c3/netket/vqs/mc/mc_state/expect.py#L133-134)

File: test/variational/test_variational.py [hunk 2]
```diff
@@ -105,6 +105,14 @@ def vstate(request):
...
     vs = nk.vqs.MCState(sa, ma, n_samples=1000, seed=SEED)
 
+    # wrap the apply function with a check that the samples have exactly 0 or 1 batch axis
+    def apply_fun_check_shape(apply_fun, params, x, *args, **kwargs):
+        assert x.ndim in (1, 2)
+        return apply_fun(params, x, *args, **kwargs)
+
+    apply_fun = vs._apply_fun
+    vs._apply_fun = partial(apply_fun_check_shape, apply_fun)
+
```
Comment (inailuig) [inline test/variational/test_variational.py:115]: let me know if this is too hacky.
I felt like it would be good to check it in many places, and here we can do so for free (compared to adding a single dedicated test)

Also, if you check for ndim==2 there is quite a few places where the ndim=1, but apparently we always wrap with to support this  https://github.com/netket/netket/blob/fe7629b996250b0661e78fb5f79e83648d1bc3c3/netket/vqs/mc/mc_state/state.py#L197-L205

File: netket/vqs/mc/kernels.py [hunk 3]
```diff
@@ -68,7 +69,7 @@ def local_value_kernel_jax(
     """
     σp, mel = O.get_conn_padded(σ)
     logpsi_σ = logpsi(pars, σ)
-    logpsi_σp = logpsi(pars, σp)
+    logpsi_σp = wrap_to_support_multiple_batch_axes(logpsi)(pars, σp)
```
Comment (PhilipVinc) [inline netket/vqs/mc/kernels.py:72]: Now that I think about it. I am not in favour of this change here. 
The `apply_fun` should be wrapped, in case, inside of the variational state.

Otherwise we end up with a complex patchwork of what supports and what does not support multiple axes. 

My idea is that those operators & co should have the simplest as possible code, and then we can put a lot of those smarts in the constructors of MCState or similar.
If a user develops/uses his own variational state, he can wrap the callable he uses as apply fun by himself. 

Again, the depths of netket's operators should stay as simple as possible..

File: netket/utils/jax.py [hunk 4]
```diff
@@ -84,3 +84,36 @@ def maybe_scalar_fun(apply_fun, pars, x, *args, **kwargs):
...
+
+def wrap_to_support_multiple_batch_axes(fun):
+    """
+    Wraps the flax-compatible apply function, assuming that the state input is the
+    second argumen with possibly many leading batch axes,
+    so that it always calls the wrapped function with a tensor with exactly 2 dimensions.
+    The result is reshaped so that it reflects the dimensions of the input.
+
+    DEVNOTE: This function is used because some users like to define their models with a vmap,
+    only supporting a single batch axis.
```
Comment (PhilipVinc) [inline netket/utils/jax.py:97]: I don't know how I feel about adding to netket 100 more lines, and a wrapper, to support a terrible habit.

We should rather instruct those 'users' in the right ways of `vectorize` their models rather than fix their bad behaviour ourself.

(I feel netket's getting too many magics...)

Comment (jwnys) [reply netket/utils/jax.py:97]: netket has always corrected for this with the numba operators. It's just bizar not to do it now for the jax operator. This 'bad' behavior has already been encouraged by all previous versions of netket.
 It just introduced easy mistakes (it took me time to understand that this is what was happening). Furthermore, all other libaries do the same: you have a single batch axis...

Comment (gcarleo) [reply netket/utils/jax.py:97]: I would be in favor of having requirements on models / user inputs rather than putting too many conversions inside of netket if possible. For example, why can't we require users to have their models take a single batch axis only?

Comment (PhilipVinc) [reply netket/utils/jax.py:97]: @jwnys a while ago we had decided to explicitly not require that models support scalar inputs, but we never made a choice on whether we would require more than 1 batch dimension or not, even though I guess my wording does indeed suggest we only require a single batch axis support, so that's the defecto standard now.

What is missing are consistent tests to catch this.

**Review by PhilipVinc — APPROVED** [review]
So, to be honest I would be more in for adding the simple reshape - reshape trick we had at the beginning rather than this 'complicated' function, for the simple reason that I believe it to be uglier but more straightforward (we could add a comment) and easier to copy-paste around.

also because we need this in a single place...

Regardless, I don't feel very strongly about this, so if @gcarleo prefers to have the wrapper function, that's fine by me

**Comment by gcarleo** [comment]
I think it is easier if we ask users to provide models that accept a single batch dimension, and error if this is not the case. I agree it can be frustrating if it is not clear that this is required of the user's model. Is there a way to add a quick test on the model input to validate it, so to speak?

**Comment by gcarleo** [comment]
> Regardless, I don't feel very strongly about this, so if @gcarleo prefers to have the wrapper function, that's fine by me

No, I don't especially like this complicated wrapper function, it's just that it would be easier to remove if we add a model validation step. Still, I think we shouldn't ask users to write models that accept 2-axes batches, it's too much of a pain for the user

**Comment by PhilipVinc** [comment]
Ok, then let's keep it this way then. Is it ok if we just go back to the original reshape - reshape?

**Comment by gcarleo** [comment]
fine... it's ugly to have these reshapes that could be summarized in a single function, but if it's commented... ok

**Review by PhilipVinc — APPROVED** [review]
